### PR TITLE
recapitate with msprime v1

### PR DESCRIPTION
**Pull request by petrelharp**
Here's a proposal for how that will work: it lets users specify an `ancestral_size`, and all populations are merged into one ancestral one of that size.

This builds on #154.

### CODE REVIEW

File: pyslim/slim_tree_sequence.py [hunk 1]
```diff
@@ -417,53 +418,67 @@ def recapitate(self,
...
+        demography = None
+        if "demography" in kwargs:
+            demography = kwargs['demography']
+        if ancestral_Ne is not None:
+            if demography is not None:
+                raise ValueError("You cannot specify both `demography` and `ancestral_Ne`.")
+            demography = msprime.Demography.from_tree_sequence(self)
+            # must set pop sizes to >0 even though we merge immediately
+            for pop in demography.populations:
+                pop.initial_size=1.0
```
Comment (petrelharp) [inline pyslim/slim_tree_sequence.py]: it seems like I shouldn't need to do this, but I do.

**Comment by petrelharp** [comment]
Could you have a look [here](https://github.com/petrelharp/pyslim/pull/1/files#diff-5197d510772af9508d710a5d4ec3471fe013a866e5641c4395c8170a63852258R36), @jeromekelleher? I'm trying to recapitate by having all pops come from a population split of an "ancestral" population. However, this doesn't work, because I get an `Attempt to sample a lineage from an inactive population` error, because the lineages I'm starting with are at time `T`, and the split is at time `T`, and so the lineages' populations are inactive at time `T`.  The only ways I see around this are:

1. have the split happen at time `T + epsilon` where epsilon is small
2. require people to set up their own ancestral demography

... but neither of these seem like great options.

**Comment by jeromekelleher** [comment]
It's not obvious to me that it's actually a good idea to do this, tbh. We're only providing this function so that the returned tree sequence is a SlimTreeSequence instance and we don't lose the reference_sequence, right? In terms of the Demography, I think there'd be less overall confusion if the user was required to specify their own. Since Slim populations don't have ``names`` it does get quite messy and confusing, though.

There will need to be some documentation somewhere about how to translate a SLiM population model into a Demography (inevitably the simple approach here won't work for many applications) so why not bite the bullet and focus on writing a documented function to do this first? So, we have a function ``make_demography(slim_ts)`` or something, that returns an appropriate demography object from a slim ts. Then, you'd have another function (or maybe this should be something that goes into the msprime API?) that migrates all lineages to a single population at the simulation start time.

Combining the definition of the demography and the simulation like it's currently done will lead to more pain in the long term, I think.

**Comment by petrelharp** [comment]
I'm starting to move towards *not* having SlimTreeSequence be a thing; that's why I wrote this as `pyslim.recapitate( )`. The goal here is to make this common and important operation be easy, so that people can do the most common, sensible thing without having to understand how to set up a demographic model in msprime. Another real good reason to have a method called "recapitate" is that people know that's what they need to do, so will go looking for it. If there just isn't a sensible deault operation, then OK, we can't do it. But, I think that what I've done here is a pretty sensible default? Let's see:

1. The most common use case is when SLiM has a single population, and they want to recapitate with a fixed Ne. 
2. The next most common is, when someone has a maybe complicated SLiM simluation, in which most of the genome is coalesced probably. The bit of history they need to simulate is remote enough in time that we don't have a real good idea of exactly what it's like, and just doing a single population with some Ne is, in my opinion, totally the way to go.

I think we definitely want to cover at least (1). Doing that just that is a bit tricky, because we'd need to go through all the roots of all the trees and check which population has lineages in, and if there's only one, changing its size to `ancestral_Ne` and simulating. Thus, the solution that I've got in the code, which covers (2) also.

> Combining the definition of the demography and the simulation like it's currently done will lead to more pain in the long term, I think.

I'm not quite sure what you mean, but one option would be to *not* allow the `demography` argument to `recapitate`, so that the `recapitate` method *only* does the default thing of coalescence-in-a-single-unstructured-population. But then the method would need to be called `recapitate_with_just_one_pop()` or something so people don't think recapitation implies a single population.  

We *are* planning to put `names` into SLiM populations soon, so if that makes things a lot easier then we could go ahead and do that.

> There will need to be some documentation somewhere about how to translate a SLiM population model into a Demography

I guess so? This could be done for a WF simulation, although not all the information (notably, population size) is in the metadata. But, it doesn't even make sense for a nonWF simulation. We do at least need an example of recapitating with "the same" WF model in the docs.

> Then, you'd have another function (or maybe this should be something that goes into the msprime API?) that migrates all lineages to a single population at the simulation start time.

Right - is there even a way to do this? I think we'd just need to add a little time epsilon, or define a whole new msprime demographic even that takes place *just after* time T (ie affects the interval (T, infty) instead of [T, infty)).

**Comment by jeromekelleher** [comment]
Fair enough on the points above. I'd need to think more about the details and maybe work through an example to give more constructive feedback, but I don't have time right now so don't block on me.

> Right - is there even a way to do this? I think we'd just need to add a little time epsilon, or define a whole new msprime demographic even that takes place just after time T (ie affects the interval (T, infty) instead of [T, infty)).

Probably easier to  use MassMigration events which will sidestep the state machine stuff entirely. You're not really modelling a population split anyway, and the state machine is only there to prevent users from making mistakes and to make the demography object easier to debug --- neither of which apply here.

**Comment by petrelharp** [comment]
> Probably easier to use MassMigration events which will sidestep the state machine stuff entirely. 

Well, these don't check for correctness, true, but they also don't actually apply, for the same reason as the population split, it turns out: note that in this example, all the nodes are in pop 0, not pop 1:
```
import tskit, msprime

ts = msprime.sim_ancestry(1, end_time=0.1, random_seed=1)

demography = msprime.Demography.from_tree_sequence(ts)
for pop in demography.populations:
    pop.initial_size=1.0
demography.add_population(name="pop_1", initial_size=1.0)
demography.add_mass_migration(
        0.1,
        source='pop_0',
        dest='pop_1',
        proportion=1.0)

ts2 = msprime.sim_ancestry(initial_state=ts, demography=demography)

print(ts2.tables.nodes)
```

**Comment by petrelharp** [comment]
I used `np.nextafter( )` to [add an epsilon to the time](https://github.com/petrelharp/pyslim/commit/d029420374f17a9770fcd482b17e14d02b754d29#diff-5197d510772af9508d710a5d4ec3471fe013a866e5641c4395c8170a63852258R70) - I don't think there's a better way to do it.

**Pull request merged**